examples/matrix_multiplication.masm

```diff
@@ -0,0 +1,735 @@
+    push.100 mem_storew dropw
+
+    # matrix B (column-major)
+    # [row=1, col=1, 0, 0]
+    # [2]
+    push.0 push.0 push.1 push.1
+    push.0 push.0 push.0 push.2
+    push.201 mem_storew dropw
+    push.200 mem_storew dropw
+
+    push.300 push.200 push.100 # [addrA, addrB, addrC]
+    exec.matrix_multiply
+
+    # matrix C (row-major)
+    # [row=1, col=1, 0, 0]
+    # [2]
+    push.0 push.0 push.1 push.1
+    push.0 push.0 push.0 push.2
+
+    padw push.301 mem_loadw
+    eqw assert dropw dropw
+    padw push.300 mem_loadw
+    eqw assert dropw dropw
+end
+
+proc.test_matrix_2_1.0
+    # matrix A (row-major)
+    # [row=1, col=4, 0, 0]
+    # [2, 3, 4, 5]
+    push.0 push.0 push.4 push.1
+    push.5 push.4 push.3 push.2
+    push.401 mem_storew dropw
+    push.400 mem_storew dropw
+
+    # matrix B (column-major)
+    # [row=4, col=1, 0, 0]
+    # [6]
+    # [7]
+    # [8]
+    # [9]
+    push.0 push.0 push.1 push.4
+    push.9 push.8 push.7 push.6
+    push.501 mem_storew dropw
+    push.500 mem_storew dropw
+
+    push.600 push.500 push.400 # [addrA, addrB, addrC]
+    exec.matrix_multiply
+
+    # matrix C (row-major)
+    # [row=1, col=1, 0, 0]
+    # [110]
+    push.0 push.0 push.1 push.1
+    push.0 push.0 push.0 push.110
+
+    padw push.601 mem_loadw
+    eqw assert dropw dropw
+    padw push.600 mem_loadw
+    eqw assert dropw dropw
+end
+
+proc.test_matrix_4_4.0
+    # matrix A (row-major)
+    # [row=2, col=4, 0, 0]
+    # [2, 3, 4, 5]
+    # [6, 7, 8, 9]
+    push.0 push.0 push.4 push.2
+    push.5 push.4 push.3 push.2
+    push.9 push.8 push.7 push.6
+    push.702 mem_storew dropw
+    push.701 mem_storew dropw
+    push.700 mem_storew dropw
+
+    # matrix B (column-major)
+    # [row=4, col=2, 0, 0]
+    # [1, 1]
+    # [1, 1]
+    # [1, 1]
+    # [1, 1]
+    push.0 push.0 push.2 push.4
+    push.1 push.1 push.1 push.1
+    push.1 push.1 push.1 push.1
+    push.802 mem_storew dropw
+    push.801 mem_storew dropw
+    push.800 mem_storew dropw
+
+    push.900 push.800 push.700 # [addrA, addrB, addrC]
+    exec.matrix_multiply
+
+    # matrix C (row-major)
+    # [row=2, col=2, 0, 0]
+    # [14, 14]
+    # [30, 30]
+    push.0 push.0 push.2 push.2
+    push.0 push.0 push.14 push.14
+    push.0 push.0 push.30 push.30
+
+    padw push.902 mem_loadw
+    eqw assert dropw dropw
+    padw push.901 mem_loadw
+    eqw assert dropw dropw
+    padw push.900 mem_loadw
+    eqw assert dropw dropw
+end
+
+proc.test_matrix_row_major_copy.0
+    push.1000
+    exec.matrix_row_major_copy
+
+    # matrix (row-major)
+    # [row=2, col=3, 0, 0]
+    # [3, 2, 0]
+    # [7, 5, 1]
+    push.0 push.0 push.3 push.2
+    push.0 push.0 push.2 push.3
+    push.0 push.1 push.5 push.7
+
+    padw push.1002 mem_loadw
+    eqw assert dropw dropw
+    padw push.1001 mem_loadw
+    eqw assert dropw dropw
+    padw push.1000 mem_loadw
+    eqw assert dropw dropw
+
+    dropw dropw dropw
+end
+
+proc.test_matrix_column_major_copy.0
```

Comment:
Testing like this is not ideal, because the later tests will have data later in the advice tape, meaning the advice tape has to be edited. A better approach would be to have tests on separate files, but for an example I think this should be good enough.

Reply:
Good point, but I agree that it's fine for this example. We can think more about better ways of testing miden assembly as we add more things to the stdlib